src/Storages/Cache/ExternalDataSourceCache.cpp

```diff
@@ -0,0 +1,224 @@
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot initialize ExternalDataSourceCache twice");
+    }
+    LOG_INFO(
+        log, "Initializing local cache for remote data sources. Local cache root path: {}, cache size limit: {}", root_dir_, limit_size_);
+    root_dir = root_dir_;
+    local_cache_bytes_read_before_flush = bytes_read_before_flush_;
+    lru_caches = std::make_unique<CacheType>(limit_size_);
+
+    /// create if root_dir not exists
+    if (!fs::exists(fs::path(root_dir)))
+    {
+        fs::create_directories(fs::path(root_dir));
+    }
+
+    recover_task_holder = context->getSchedulePool().createTask("recover local cache metadata for remote files", [this]{ recoverTask(); });
+    recover_task_holder->activateAndSchedule();
+}
+
+String ExternalDataSourceCache::calculateLocalPath(IRemoteFileMetadataPtr metadata) const
+{
+    // add version into the full_path, and not block to read the new version
+    String full_path = metadata->getName() + ":" + metadata->remote_path
+        + ":" + metadata->getVersion();
+    UInt128 hashcode = sipHash128(full_path.c_str(), full_path.size());
+    String hashcode_str = getHexUIntLowercase(hashcode);
+    return fs::path(root_dir) / hashcode_str.substr(0, 3) / hashcode_str;
+}
+
+std::pair<RemoteCacheControllerPtr, std::unique_ptr<ReadBuffer>>
+ExternalDataSourceCache::createReader(ContextPtr context, IRemoteFileMetadataPtr remote_file_metadata, std::unique_ptr<ReadBuffer> & read_buffer)
+{
+    // If something is wrong on startup, rollback to read from the original ReadBuffer
+    if (!isInitialized())
+    {
+        LOG_ERROR(log, "ExternalDataSourceCache has not been initialized");
+        return {nullptr, std::move(read_buffer)};
+    }
+
+    auto remote_path = remote_file_metadata->remote_path;
+    const auto & last_modification_timestamp = remote_file_metadata->last_modification_timestamp;
+    auto local_path = calculateLocalPath(remote_file_metadata);
+    std::lock_guard lock(mutex);
+    auto cache = lru_caches->get(local_path);
+    if (cache)
+    {
+        // the remote file has been updated, need to redownload
+        if (!cache->isValid() || cache->isModified(remote_file_metadata))
+        {
+            LOG_TRACE(
+                log,
+                "Remote file ({}) has been updated. Last saved modification time: {}, actual last modification time: {}",
+                remote_path,
+                std::to_string(cache->getLastModificationTimestamp()),
+                std::to_string(last_modification_timestamp));
+            cache->markInvalid();
+        }
+        else
+        {
+            return {cache, nullptr};
+        }
+    }
+
+    if (!fs::exists(local_path))
+        fs::create_directories(local_path);
+
+    // cache is not found or is invalid
+    auto new_cache = std::make_shared<RemoteCacheController>(remote_file_metadata, local_path, local_cache_bytes_read_before_flush);
+    if (!lru_caches->trySet(local_path, new_cache))
```

Comment:
Looks like here can be a race between putting value in cache (at this moment `canRelease` == true), and making it unreleasable. So it can be evicted before we call `allocFile` where increment ref count.

Reply:
fixed